packages/commons/src/repositories/ReadModelRepository.ts

```diff
@@ -25,6 +25,72 @@ export type Collections =
   | TenantCollection
   | AttributeCollection;
 
+type BuildQueryKey<TPrefix extends string, TKey> = `${TPrefix}.${TKey &
+  string}`;
+type BuilIndexedQueryKey<TPrefix extends string, TKey> = `${TPrefix}.${TKey &
+  string}.${number}`;
+
+type ExtractQueryKeysFromRecord<
+  T extends Record<string, unknown>,
+  TPrefix extends string
+> = {
+  [TKey in keyof T]: T[TKey] extends Record<string, unknown>
+    ? MongoQueryKeys<T[TKey], BuildQueryKey<TPrefix, TKey>>
+    : T[TKey] extends unknown[]
+    ? MongoQueryKeys<
+        T[TKey],
+        BuildQueryKey<TPrefix, TKey> | BuilIndexedQueryKey<TPrefix, TKey>
+      >
+    : BuildQueryKey<TPrefix, TKey>;
+}[keyof T];
+
+/**
+ * Extracts recursively all the possible document db query keys from an object
+ *
+ * @example
+ * type Test = {
+ *   a: string;
+ *   b: {
+ *     c: string;
+ *   }
+ *   d: {
+ *     e: string;
+ *   }[]
+ * }
+ *
+ * type Result = MongoQueryKeys<Test>;
+ * //      ^ "data.a" | "data.b" | "data.b.c" | "data.d" | "data.d.e"
+ */
+type MongoQueryKeys<T, TPrefix extends string = "data"> = NonNullable<
+  | TPrefix
+  | (T extends unknown[]
+      ? MongoQueryKeys<T[number], TPrefix>
+      : T extends Record<string, unknown>
+      ? ExtractQueryKeysFromRecord<T, TPrefix>
+      : never)
+>;
+
+/**
+ * Type of the filter that can be used to query the read model.
+ * It extends the mongodb filter type by adding all the possible model query keys.
+ */
+export type ReadModelFilter<TSchema> = {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  [P in MongoQueryKeys<WithId<{ data: TSchema }["data"]>>]?: any; // We should find a better typing here!
+} & {
+  $and?: Array<Filter<{ data: TSchema }>>;
```

Comment:
I'm still not convinced about replicating this part of the typing

Reply:
@galales I made some changes, let me know what you think!

Reply:
if we don't lose expressivity, then it's better.
My concern is about replicating types that are (are they?) in the library.
But it's not a major issue, we can use this implementation 👍